tls/s2n_config.c

```diff
@@ -562,6 +571,50 @@ static int s2n_config_add_cert_chain_and_key_impl(struct s2n_config *config, str
     return S2N_SUCCESS;
 }
 
+S2N_RESULT s2n_config_validate_certificate_preferences(const struct s2n_config *config,
+        const struct s2n_security_policy *security_policy)
+{
+    RESULT_ENSURE_REF(config);
+    RESULT_ENSURE_REF(security_policy);
+    if (!security_policy->certificate_preferences_apply_locally
+            || security_policy->certificate_signature_preferences == NULL) {
+        return S2N_RESULT_OK;
+    }
```

Comment:
Not sure if it's worth removing the optimization, but I feel like it would make more sense for this check to be in `s2n_security_policy_validate_certificate_chain`. It seems a bit strange for validate_certificate_chain to fail even when `certificate_preferences_apply_locally` is false. And `s2n_config_validate_certificate_preferences` shouldn't really care about these fields since it calls validate_certificate_chain.

Reply:
Let me go ahead and benchmark it. I think the point about moving the null check makes sense, so as long as there isn't a big impact on the performance I'll go ahead and make the change.

Reply:
It only saves about 15 nanoseconds, so I'll go ahead and remove it. But ran into some other issues with that, which are addressed in #4416